x-pack/platform/plugins/shared/cases/public/components/markdown_editor/use_image_paste_upload.tsx

```diff
@@ -0,0 +1,245 @@
+import { createUploadState, type UploadState } from '@kbn/shared-ux-file-upload/src/upload_state';
+import { useUploadDone } from '../files/use_upload_done';
+import type { MarkdownEditorRef } from './types';
+import { NO_SIMULTANEOUS_UPLOADS_MESSAGE, UNSUPPORTED_MIME_TYPE_MESSAGE } from './translations';
+import { SUPPORTED_PASTE_MIME_TYPES } from './constants';
+import { constructFileKindIdByOwner } from '../../../common/files';
+import { type Owner } from '../../../common/constants/types';
+import { OWNERS } from '../../../common/constants';
+
+interface UseImagePasteUploadArgs {
+  editorRef: React.ForwardedRef<MarkdownEditorRef | null>;
+  field: FieldHook<string>;
+  caseId: string;
+  owner: string;
+  fileKindId: string;
+  setErrors: React.Dispatch<React.SetStateAction<Array<string | Error>>>;
+}
+
+/**
+ * Returns a placeholder string for the file that is being uploaded
+ */
+function generatePlaceholderCopy(filename: string, extension?: string) {
+  return `<!-- uploading "${filename}${extension ? `.${extension}` : ''}" -->`;
+}
+
+/**
+ * Returns a markdown link for the file with a link to the asset
+ */
+export const markdownImage = (fileName: string, fileUrl: string, ext?: string) =>
+  `![${fileName}${ext ? `.${ext}` : ''}](${fileUrl})`;
+
+function getTextarea(editorRef: React.ForwardedRef<MarkdownEditorRef | null>) {
+  if (!editorRef || typeof editorRef === 'function' || !editorRef.current) {
+    return null;
+  }
+  return editorRef.current.textarea;
+}
+
+const validOwners = new Set<Owner>(OWNERS);
+const isOwner = (o: string): o is Owner => validOwners.has(o as Owner);
+
+export function useImagePasteUpload({
+  editorRef,
+  field,
+  caseId,
+  owner,
+  fileKindId,
+  setErrors,
+}: UseImagePasteUploadArgs): {
+  isUploading: boolean;
+  uploadState: UploadState;
+} {
+  const { client } = useFilesContext();
+  const kind = client.getFileKind(fileKindId);
+
+  const uploadState = useMemo(
+    () =>
+      createUploadState({
+        client,
+        fileKind: kind,
+        allowRepeatedUploads: false,
+      }),
+    [client, kind]
+  );
+
+  const [isUploading, setIsUploading] = useState(false);
+  const [uploadPlaceholder, setUploadPlaceholder] = useState<string | null>(null);
+  const [uploadingFileName, setUploadingFileName] = useState<string | null>(null);
+  const textarea = getTextarea(editorRef);
+
+  const replacePlaceholder = useCallback(
+    (file: DoneNotification) => {
+      if (!textarea) return;
+      if (!isOwner(owner)) return;
+
+      const newText = textarea.value.replace(
+        uploadPlaceholder ?? '',
+        markdownImage(
+          file.fileJSON.name,
+          client.getDownloadHref({
+            id: file.id,
+            fileKind: constructFileKindIdByOwner(owner),
+          }),
+          file.fileJSON.extension
+        )
+      );
+      field.setValue(newText);
+    },
+    [textarea, uploadPlaceholder, field, client, owner]
+  );
+  const onDone = useUploadDone({
```

Comment:
Can you elaborate on why the done check lives in its own hook?

Reply:
File is removed in latest commits.